azurelinuxagent/common/event.py

```diff
@@ -224,13 +234,15 @@ def save_event(self, data):
             raise EventError(msg)
 
         existing_events = os.listdir(self.event_dir)
-        if len(existing_events) >= 1000:
-            existing_events.sort()
-            oldest_files = existing_events[:-999]
-            logger.warn("Too many files under: {0}, removing oldest".format(self.event_dir))
+        if len(existing_events) >= MAX_NUMBER_OF_EVENTS:
+            logger.periodic_warn(logger.EVERY_MINUTE, "[PERIODIC] Too many files under: {0}, current count:  {1}, "
+                                                      "removing oldest".format(self.event_dir, len(existing_events)))
             try:
-                for f in oldest_files:
-                    os.remove(os.path.join(self.event_dir, f))
+                existing_events = os.listdir(self.event_dir)
```

Comment:
When this test does `add_event`, it was trying to add a new event, but the event already had 1000 events in the file, because of which it generated a `log.warn` - which again generates and adds one more event, and in the process deletes the oldest file.

Now, the earlier `add_event` call continues and does the os.remove of the oldest file, but it got deleted earlier itself, and thus hits the following issue [TravisLink](https://travis-ci.org/Azure/WALinuxAgent/jobs/614418917):
Eg:
```
======================================================================
ERROR: test_telemetry_logger_verify_maximum_recursion_depths_doesnt_happen (tests.common.test_logger.TestLogger)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/travis/build/Azure/WALinuxAgent/tests/common/test_logger.py", line 354, in test_telemetry_logger_verify_maximum_recursion_depths_doesnt_happen
    logger.warn('Test Log - {0} - 1 - Warning'.format(i))
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/logger.py", line 289, in warn
    DEFAULT_LOGGER.warn(msg_format, *args)
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/logger.py", line 79, in warn
    self.log(LogLevel.WARNING, msg_format, *args)
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/logger.py", line 140, in log
    write_log(appender)
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/logger.py", line 117, in write_log
    log_appender.write(level, log_item)
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/logger.py", line 207, in write
    self.event_func(level, msg)
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/event.py", line 510, in add_log_event
    reporter.add_log_event(level, message)
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/event.py", line 314, in add_log_event
    self.save_event(json.dumps(data))
  File "/home/travis/build/Azure/WALinuxAgent/azurelinuxagent/common/event.py", line 244, in save_event
    os.remove(os.path.join(self.event_dir, event_file))
OSError: [Errno 2] No such file or directory: '/tmp/TestLogger_ARDFnT/1574239919504255.tld'
```

This was specifically seen in Python2 and could be due to the event order in the list, and being picked after sorting.

FYI - @larohra

Reply:
So it turns out I didn't need the new list but understand when the sort is called. Once the sort is called later, it fixes the deletion problem.